File: typescript/packages/mechanisms/evm/src/batch-settlement/storage/fileStorage.ts
import { mkdir, open, readdir, readFile, stat, unlink, writeFile } from "node:fs/promises";
import { constants } from "node:fs";
import { dirname, join } from "node:path";
import { randomUUID } from "node:crypto";

import { isNodeEnoent, readJsonFile, resolveWithinDir, writeJsonAtomic } from "../storage-utils";
import { normalizeChannelId } from "../utils";
import type { FileChannelStorageOptions } from "../types";
import type { ChannelLockStorage, ChannelStorage, Channel, ChannelUpdateResult } from "./channel";

export type { FileChannelStorageOptions };

const FILE_LOCK_MAX_ATTEMPTS = 50;
const FILE_LOCK_RETRY_INTERVAL_MS = 10;
const FILE_LOCK_STALE_MS = 30_000;

/**
 * Owner record written into every exclusive lock marker.
 *
 * The on-disk protocol is shared with the Go SDK: a lock file contains JSON
 * `{ pid, token, createdAt }`. Stealing is only allowed when the owner is
 * provably gone (dead PID) or when an unreadable legacy marker is older than
 * `staleMs`. A live owner is never stolen, no matter how old the marker is.
 */
export type FileLockOwner = {
  pid: number;
  token: string;
  createdAt: number;
};

export type AcquireExclusiveFileOptions = {
  maxAttempts?: number;
  retryIntervalMs?: number;
  staleMs?: number;
};

/**
 * Node.js file-backed {@link ChannelStorage} for batch-settlement channels.
 */
export class FileChannelStorage<T extends Channel = Channel>
  implements ChannelStorage<T>, ChannelLockStorage
{
  private readonly root: string;

  /**
   * Creates file-backed channel storage under the given root directory.
   *
   * @param options - Configuration including the storage root directory.
   */
  constructor(options: FileChannelStorageOptions) {
    this.root = options.directory;
  }

  /**
   * Loads a persisted channel record, if present.
   *
   * @param channelId - The channel identifier (path segment is lowercased).
   * @returns Parsed channel record or `undefined` when the file is missing.
   */
  async get(channelId: string): Promise<T | undefined> {
    return readJsonFile<T>(this.filePath(channelId));
  }

  /**
   * Lists all stored channel records by reading the server directory.
   *
   * @returns Channel records sorted by channelId; empty array if the directory is missing.
   */
  async list(): Promise<T[]> {
    const dir = join(this.root, "server");
    let names: string[];
    try {
      names = await readdir(dir);
    } catch (err: unknown) {
      if (isNodeEnoent(err)) return [];
      throw err;
    }

    const channels: T[] = [];
    for (const name of names) {
      if (!name.endsWith(".json")) continue;
      const path = join(dir, name);
      try {
        const raw = await readFile(path, "utf8");
        channels.push(JSON.parse(raw) as T);
      } catch (err: unknown) {
        // Skip files that disappeared between readdir and readFile (e.g. concurrent delete).
        // Rethrow other failures (corrupt JSON, permission denied) so callers see them.
        if (isNodeEnoent(err)) continue;
        throw err;
      }
    }
    return channels.sort((a, b) => a.channelId.localeCompare(b.channelId));
  }

  /**
   * Atomically inspects and mutates a channel record under a cross-process file lock.
   *
   * @param channelId - The channel identifier.
   * @param update - Mutation callback. Return `undefined` to delete, or `current` to leave unchanged.
   * @returns The final stored channel and whether storage updated, stayed unchanged, or deleted.
   */
  async updateChannel(
    channelId: string,
    update: (current: T | undefined) => T | undefined,
  ): Promise<ChannelUpdateResult<T>> {
    const lockPath = this.filePath(channelId) + ".lock";
    await mkdir(dirname(lockPath), { recursive: true });
    const lockHandle = await acquireExclusiveFile(lockPath);

    try {
      const path = this.filePath(channelId);
      let current: T | undefined;
      try {
        const raw = await readFile(path, "utf8");
        current = JSON.parse(raw) as T;
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }

      const next = update(current);
      if (next === current) {
        return { channel: current, status: "unchanged" };
      }

      if (!next) {
        try {
          await unlink(path);
        } catch (err: unknown) {
          if (!isNodeEnoent(err)) throw err;
        }
        await this.dropHold(channelId);
        return { channel: undefined, status: current ? "deleted" : "unchanged" };
      }

      await writeJsonAtomic(path, next);
      return { channel: next, status: "updated" };
    } finally {
      await lockHandle.close();
      await unlink(lockPath).catch(() => {});
    }
  }

  /**
   * Acquires a per-channel admission lock via a sidecar hold file.
   *
   * Serialized with {@link FileChannelStorage.release} and {@link FileChannelStorage.isHeld}
   * on `{id}.hold.lock` so an expired hold cannot be unlinked out from under a new holder.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   * @param ttlMs - Lock time-to-live in milliseconds.
   * @returns Whether this request now holds the lock.
   */
  async acquire(channelId: string, pendingId: string, ttlMs: number): Promise<boolean> {
    return this.withHoldLock(channelId, async () => {
      const path = this.holdPath(channelId);
      await mkdir(dirname(path), { recursive: true });
      try {
        const existing = JSON.parse(await readFile(path, "utf8")) as {
          pendingId: string;
          expiresAt: number;
        };
        if (existing.expiresAt > Date.now()) {
          return false;
        }
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
      await writeFile(path, JSON.stringify({ pendingId, expiresAt: Date.now() + ttlMs }), "utf8");
      return true;
    });
  }

  /**
   * Releases the admission lock only when `pendingId` still holds it.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   */
  async release(channelId: string, pendingId: string): Promise<void> {
    await this.withHoldLock(channelId, async () => {
      const path = this.holdPath(channelId);
      try {
        const hold = JSON.parse(await readFile(path, "utf8")) as { pendingId: string };
        if (hold.pendingId !== pendingId) return;
        await unlink(path);
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
    });
  }

  /**
   * Returns whether a live admission lock exists, optionally matching `pendingId`.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - When set, require this request to hold the lock.
   * @returns Whether a live lock (or this request's lock) is present.
   */
  async isHeld(channelId: string, pendingId?: string): Promise<boolean> {
    return this.withHoldLock(channelId, async () => {
      try {
        const hold = JSON.parse(await readFile(this.holdPath(channelId), "utf8")) as {
          pendingId: string;
          expiresAt: number;
        };
        if (hold.expiresAt <= Date.now()) return false;
        return pendingId === undefined || hold.pendingId === pendingId;
      } catch (err: unknown) {
        if (isNodeEnoent(err)) return false;
        throw err;
      }
    });
  }

  /**
   * Absolute path to the JSON file for a channel.
   *
   * @param channelId - The channel identifier.
   * @returns Filesystem path under `{root}/server/...`.
   * @throws When `channelId` is not a canonical `bytes32` string or escapes the storage root.
   */
  private filePath(channelId: string): string {
    const id = normalizeChannelId(channelId);
    return resolveWithinDir(join(this.root, "server"), `${id}.json`);
  }

  /**
   * Absolute path to the admission hold sidecar for a channel.
   *
   * @param channelId - The channel identifier.
   * @returns Filesystem path under `{root}/server/{id}.hold`.
   */
  private holdPath(channelId: string): string {
    const id = normalizeChannelId(channelId);
    return resolveWithinDir(join(this.root, "server"), `${id}.hold`);
  }

  /**
   * Drops the admission hold for a deleted channel row.
   *
   * @param channelId - The channel identifier.
   */
  private async dropHold(channelId: string): Promise<void> {
    await this.withHoldLock(channelId, async () => {
      try {
        await unlink(this.holdPath(channelId));
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
    });
  }

  /**
   * Serializes acquire, release, and isHeld on the `.hold` sidecar.
   *
   * @param channelId - The channel identifier.
   * @param fn - Work to run while holding `{id}.hold.lock`.
   * @returns The resolved result of `fn`.
   */
  private async withHoldLock<R>(channelId: string, fn: () => Promise<R>): Promise<R> {
    const lockPath = this.holdPath(channelId) + ".lock";
    await mkdir(dirname(lockPath), { recursive: true });
    const lockHandle = await acquireExclusiveFile(lockPath);
    try {
      return await fn();
    } finally {
      await lockHandle.close();
      await unlink(lockPath).catch(() => {});
    }
  }
}

/**
 * Creates an exclusive lock file with an owner marker.
 *
 * The marker holds JSON `{ pid, token, createdAt }` under the same on-disk
 * protocol as the Go SDK. A contended marker is unlinked only when its owner
 * is provably gone (dead PID) or when a legacy/unreadable marker is older
 * than `staleMs`. A live owner is never stolen: slow writers keep the lock
 * and contenders fail with `contended` after `maxAttempts`.
 *
 * @param lockPath - Absolute path for the lock file (created with `O_EXCL`).
 * @param options - Attempt, retry, and stale-mtime bounds.
 * @returns Writable file handle for the lock file; caller must close it to release.
 * @throws When the lock remains contended after `maxAttempts`.
 */
export async function acquireExclusiveFile(
  lockPath: string,
  options: AcquireExclusiveFileOptions = {},
) {
  const maxAttempts = options.maxAttempts ?? FILE_LOCK_MAX_ATTEMPTS;
  const retryIntervalMs = options.retryIntervalMs ?? FILE_LOCK_RETRY_INTERVAL_MS;
  const staleMs = options.staleMs ?? FILE_LOCK_STALE_MS;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      const handle = await open(
        lockPath,
        constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY,
      );
      try {
        await handle.writeFile(
          JSON.stringify({ pid: process.pid, token: randomUUID(), createdAt: Date.now() }),
        );
        await handle.sync();
      } catch {
        // Best effort: the O_EXCL marker already serializes writers even if the
        // owner payload cannot be persisted; steal decisions fall back to mtime.
      }
      return handle;
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") {
        throw err;
      }
      if (await isStaleLock(lockPath, staleMs)) {
        await unlink(lockPath).catch(() => {});
        continue;
      }
      await new Promise(resolve => setTimeout(resolve, retryIntervalMs));
    }
  }
  throw new Error(`acquire lock ${lockPath}: contended`);
}

/**
 * Returns whether `lockPath` may be unlinked and retried.
 *
 * A marker with a readable owner is stale only when its PID is provably gone.
 * Legacy or unreadable markers (empty file from an older SDK, partial write
 * after a crash) fall back to mtime age so crash debris cannot pin the channel
 * forever.
 *
 * @param lockPath - Absolute path for the lock file.
 * @param staleMs - Age after which a legacy marker is treated as crash debris.
 * @returns Whether the caller may unlink and retry exclusive create.
 */
async function isStaleLock(lockPath: string, staleMs: number): Promise<boolean> {
  let mtimeMs = 0;
  try {
    mtimeMs = (await stat(lockPath)).mtimeMs;
  } catch (err: unknown) {
    if (isNodeEnoent(err)) return true;
    throw err;
  }
  let raw = "";
  try {
    raw = await readFile(lockPath, "utf8");
  } catch (err: unknown) {
    if (isNodeEnoent(err)) return true;
    throw err;
  }
  if (!raw.trim()) {
    return Date.now() - mtimeMs >= staleMs;
  }
  try {
    const owner = JSON.parse(raw) as Partial<FileLockOwner>;
    if (typeof owner.pid !== "number") {
      return Date.now() - mtimeMs >= staleMs;
    }
    return !isOwnerAlive(owner.pid);
  } catch {
    return Date.now() - mtimeMs >= staleMs;
  }
}

/**
 * Reports whether a lock-owner PID is still alive.
 *
 * Uses signal `0`: `ESRCH` means provably gone, `EPERM` means alive but
 * un-signallable, and any other failure fails closed to alive so a live writer
 * is never stolen.
 *
 * @param pid - PID recorded in the lock marker.
 * @returns Whether the owner process may still be running.
 */
function isOwnerAlive(pid: number): boolean {
  if (!Number.isInteger(pid) || pid <= 0) {
    return false;
  }
  if (pid === process.pid) {
    return true;
  }
  try {
    process.kill(pid, 0);
    return true;
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException | undefined)?.code;
    if (code === "ESRCH") {
      return false;
    }
    return true;
  }
}

File: go/mechanisms/evm/batch-settlement/storage/file_storage.go
package storage

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
)

const (
	fileLockMaxAttempts     = 50
	fileLockRetryIntervalMs = 10
	fileLockStaleMs         = 2_000
)

// FileChannelStorage is a file-backed ChannelStorage. Each record is stored as
// {root}/server/{channelId}.json. UpdateChannel is serialised through an
// exclusive lock file ({channelId}.json.lock) so concurrent writers cannot
// interleave.
type FileChannelStorage[T ChannelRecord[T]] struct {
	root string
}

var (
	_ ChannelStorage[*Channel] = (*FileChannelStorage[*Channel])(nil)
	_ ChannelLockStorage       = (*FileChannelStorage[*Channel])(nil)
)

// NewFileChannelStorage returns a file-backed channel store.
func NewFileChannelStorage[T ChannelRecord[T]](opts batchsettlement.FileChannelStorageOptions) *FileChannelStorage[T] {
	return &FileChannelStorage[T]{root: opts.Directory}
}

func (s *FileChannelStorage[T]) filePath(channelId string) (string, error) {
	id, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return "", err
	}
	return batchsettlement.ResolveWithinDir(filepath.Join(s.root, "server"), id+".json")
}

func (s *FileChannelStorage[T]) holdPath(channelId string) (string, error) {
	id, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return "", err
	}
	return batchsettlement.ResolveWithinDir(filepath.Join(s.root, "server"), id+".hold")
}

// Get loads a persisted channel record, or the zero T when the file is missing.
func (s *FileChannelStorage[T]) Get(channelId string) (T, error) {
	var zero T
	path, err := s.filePath(channelId)
	if err != nil {
		return zero, err
	}
	var out T
	ok, err := batchsettlement.ReadJSONFile(path, &out)
	if err != nil {
		return zero, err
	}
	if !ok {
		return zero, nil
	}
	return out, nil
}

// Set writes session as JSON under the canonical channel id.
func (s *FileChannelStorage[T]) Set(channelId string, session T) error {
	path, err := s.filePath(channelId)
	if err != nil {
		return err
	}
	return batchsettlement.WriteJSONAtomic(path, session)
}

// Delete removes the channel file and its admission hold sidecar.
func (s *FileChannelStorage[T]) Delete(channelId string) error {
	path, err := s.filePath(channelId)
	if err != nil {
		return err
	}
	if err := os.Remove(path); err != nil && !batchsettlement.IsNotExist(err) {
		return err
	}
	return s.dropHold(channelId)
}

// List returns stored records sorted by channelId.
func (s *FileChannelStorage[T]) List() ([]T, error) {
	dir := filepath.Join(s.root, "server")
	entries, err := os.ReadDir(dir)
	if err != nil {
		if batchsettlement.IsNotExist(err) {
			return []T{}, nil
		}
		return nil, err
	}

	sessions := make([]T, 0, len(entries))
	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasSuffix(name, ".json") || strings.HasSuffix(name, ".lock") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			if batchsettlement.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		var out T
		if err := json.Unmarshal(raw, &out); err != nil {
			return nil, fmt.Errorf("unmarshal %s: %w", name, err)
		}
		sessions = append(sessions, out)
	}
	sort.Slice(sessions, func(i, j int) bool { return sessions[i].Base().ChannelId < sessions[j].Base().ChannelId })
	return sessions, nil
}

// UpdateChannel atomically reads, mutates, and writes a channel record under an
// exclusive lock file. Returning a different pointer commits the new session;
// returning the zero T deletes the file; returning the same pointer is a no-op.
func (s *FileChannelStorage[T]) UpdateChannel(channelId string, update func(current T) T) (*ChannelUpdateResult[T], error) {
	path, err := s.filePath(channelId)
	if err != nil {
		return nil, err
	}
	lockPath := path + ".lock"

	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return nil, fmt.Errorf("mkdir %s: %w", filepath.Dir(lockPath), err)
	}

	lockFile, err := AcquireExclusiveFile(lockPath, nil)
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = lockFile.Close()
		_ = os.Remove(lockPath)
	}()

	var current T
	ok, err := batchsettlement.ReadJSONFile(path, &current)
	if err != nil {
		return nil, err
	}
	if !ok {
		current = zeroRecord[T]()
	}

	next := update(current)
	if sameRecord(next, current) {
		return &ChannelUpdateResult[T]{Channel: current, Status: ChannelUnchanged}, nil
	}
	if isZeroRecord(next) {
		if !ok {
			return &ChannelUpdateResult[T]{Status: ChannelUnchanged}, nil
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return nil, rmErr
		}
		if dropErr := s.dropHold(channelId); dropErr != nil {
			return nil, dropErr
		}
		return &ChannelUpdateResult[T]{Status: ChannelDeleted}, nil
	}
	if err := batchsettlement.WriteJSONAtomic(path, next); err != nil {
		return nil, err
	}
	return &ChannelUpdateResult[T]{Channel: next, Status: ChannelUpdated}, nil
}

// Acquire takes a per-channel admission lock via a sidecar hold file.
//
// Serialized with Release and IsHeld on {id}.hold.lock so an expired hold
// cannot be unlinked out from under a new holder. Not re-entrant: a live hold,
// including one owned by the same pendingId, is a miss.
func (s *FileChannelStorage[T]) Acquire(channelId string, pendingId string, ttlMs int64) (bool, error) {
	var acquired bool
	err := s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return fmt.Errorf("mkdir %s: %w", filepath.Dir(path), err)
		}
		raw, readErr := os.ReadFile(path)
		if readErr != nil && !batchsettlement.IsNotExist(readErr) {
			return readErr
		}
		if readErr == nil {
			var existing admissionLock
			if unmarshalErr := json.Unmarshal(raw, &existing); unmarshalErr != nil {
				return unmarshalErr
			}
			if existing.ExpiresAt > time.Now().UnixMilli() {
				acquired = false
				return nil
			}
		}
		record, err := json.Marshal(admissionLock{PendingId: pendingId, ExpiresAt: time.Now().UnixMilli() + ttlMs})
		if err != nil {
			return err
		}
		if err := os.WriteFile(path, record, 0o644); err != nil {
			return err
		}
		acquired = true
		return nil
	})
	return acquired, err
}

// Release drops the admission lock only when pendingId still holds it.
func (s *FileChannelStorage[T]) Release(channelId string, pendingId string) error {
	return s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			if batchsettlement.IsNotExist(err) {
				return nil
			}
			return err
		}
		var hold admissionLock
		if err := json.Unmarshal(raw, &hold); err != nil {
			return err
		}
		if hold.PendingId != pendingId {
			return nil
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return rmErr
		}
		return nil
	})
}

// IsHeld reports whether a live admission lock exists, optionally matching pendingId.
func (s *FileChannelStorage[T]) IsHeld(channelId string, pendingId string) (bool, error) {
	var held bool
	err := s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			if batchsettlement.IsNotExist(err) {
				held = false
				return nil
			}
			return err
		}
		var hold admissionLock
		if err := json.Unmarshal(raw, &hold); err != nil {
			return err
		}
		if hold.ExpiresAt <= time.Now().UnixMilli() {
			held = false
			return nil
		}
		if pendingId == "" {
			held = true
			return nil
		}
		held = hold.PendingId == pendingId
		return nil
	})
	return held, err
}

func (s *FileChannelStorage[T]) dropHold(channelId string) error {
	return s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return rmErr
		}
		return nil
	})
}

func (s *FileChannelStorage[T]) withHoldLock(channelId string, fn func() error) error {
	path, err := s.holdPath(channelId)
	if err != nil {
		return err
	}
	lockPath := path + ".lock"
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return fmt.Errorf("mkdir %s: %w", filepath.Dir(lockPath), err)
	}
	lockFile, err := AcquireExclusiveFile(lockPath, nil)
	if err != nil {
		return err
	}
	defer func() {
		_ = lockFile.Close()
		_ = os.Remove(lockPath)
	}()
	return fn()
}

// ExclusiveFileOptions bounds exclusive lock-file create attempts.
type ExclusiveFileOptions struct {
	MaxAttempts     int
	RetryIntervalMs int
	StaleMs         int
}

func (o *ExclusiveFileOptions) withDefaults() ExclusiveFileOptions {
	out := ExclusiveFileOptions{}
	if o != nil {
		out = *o
	}
	if out.MaxAttempts <= 0 {
		out.MaxAttempts = fileLockMaxAttempts
	}
	if out.RetryIntervalMs <= 0 {
		out.RetryIntervalMs = fileLockRetryIntervalMs
	}
	if out.StaleMs <= 0 {
		out.StaleMs = fileLockStaleMs
	}
	return out
}

// AcquireExclusiveFile creates lockPath with O_EXCL, polling until the marker
// is free or stale (mtime older than StaleMs). Stale markers are unlinked so a
// crash cannot pin the channel forever.
func AcquireExclusiveFile(lockPath string, opts *ExclusiveFileOptions) (*os.File, error) {
	cfg := opts.withDefaults()
	staleAfter := time.Duration(cfg.StaleMs) * time.Millisecond
	retryInterval := time.Duration(cfg.RetryIntervalMs) * time.Millisecond
	for attempt := 0; attempt < cfg.MaxAttempts; attempt++ {
		f, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
		if err == nil {
			return f, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return nil, fmt.Errorf("acquire lock %s: %w", lockPath, err)
		}
		info, statErr := os.Stat(lockPath)
		if statErr != nil {
			if batchsettlement.IsNotExist(statErr) {
				continue
			}
			return nil, fmt.Errorf("acquire lock %s: %w", lockPath, statErr)
		}
		if time.Since(info.ModTime()) >= staleAfter {
			_ = os.Remove(lockPath)
			continue
		}
		time.Sleep(retryInterval)
	}
	return nil, fmt.Errorf("acquire lock %s: contended", lockPath)
}
